### ISSUE TITLE
"Property is missing" error when mapping over array of disjoint union type

### ISSUE THREAD

**Issue by Mr-Wallet** [facebook/flow#5876]
```javascript
const x = { a: 1 };
const y = { b: 2 };

const thingsToGet = [
  [x, 'a'],
  [y, 'b']
];

thingsToGet.map(([obj, prop]) => {
  return obj[prop];
});
```
The following results in an error because  `'b'` is not found on `x`, and `'a'` is not found on `y`. However there is clearly no attempt to access such properties.
I attempted to type `thingsToGet` with the following:
`type thingsType = ([typeof x, $Keys<typeof x>]|[typeof y, $Keys<typeof y>])[];`
This explicitly states that the keys of `y` will never be mixed with the object `x` and vice versa. Flow does not complain about the declaration of `thingsToGet` so presumably it believes the typing. And yet it still errors on the `map` for the same reason! If it believes the explicit typing, it should also believe that the map operation is safe.

[Try Flow link](https://flow.org/try/#0MYewdgzgLgBAHjAvDA3jAhgLhgRhgXwG4AoUSWATyVRgCNsAmAk4gelZgBUKAHASzABzGFAAWAwRE4gA4gFNYAdz5iRouTABmIADY6QyoTABOciAFcdUCDAFqNEdAFsNc48ZDG2HKLw1iJKT9qAAoAbV8eORBNeAAaGAASAGk5CggAHkjo2LgAPgBdAB8IvxiYCgSUtMzs8opCgEowgpIyaDVA6XlYZDDiGBgwuASAcnRRgriBocqYUdpJ4lbiYgChKVkFADondB4Q8JBaACsEng8eAsakPNQZ0yhzYzAYY5Owi5ArknxGwiAA)

**Comment by SamChou19815** [comment]
This is expected behavior. This is too dynamic

**Comment by SamChou19815** [comment]
Yes, because we don't support this kind of distribute over union.
Here is a reduced example: https://flow.org/try/#1N4Igxg9gdgZglgcxALlAIwIZoKYBsD6uEEAztvhgE6UYCe+JADpdhgCYowa5kA0I2KAFcAtiRQAXSkOz9sADwxgJ+NPTbYuQ3BMnTZA+Y2yU4IwRO4A6SFBIrGVDGM7c+h46fNRLuKxJIGWh8MeT0ZfhYlCStpHzNsFBAMIQkIEQwJODAQfiEyfBE4eWw2fDgofDBMsAALfAA3KjgsXGxxZC4eAw0G-GhcWn9aY3wWZldu-g1mbGqJUoBaCRHEzrcDEgBrbAk62kXhXFxJ923d-cPRHEpTgyEoMDaqZdW7vKgoOfaSKgOKpqmDA+d4gB5fMA-P6LCCMLLQbiLOoYCqgh6-GDYRYIXYLSgkRZkCR4jpddwPfJLZjpOBkO4AX34kA0SVs9gABPJ2QBednAdkYZDsgCM7PpAG4ADqPaAc2g8vnstBCgBMYqlUGlAAoANoQNAAK147OpjAAukKdStjBAYJzjQASADS2FoJAAPNbsLbOQA+M3sgA+7Ktqx9tEdLrdnrDdto-oAlDzfXzpez2SwJEJKFB2fqDTrTWaNRLpbkQA0TCQ4NAkg0AAxWFUANgAzFZ6yB6UA
TS has the same limitation: https://www.typescriptlang.org/play/?#code/MYewdgzgLgBAHjAvDA3jAhgLhgRhgXwG4AoUSWATyVRgCNsAmAk4gCgG0RaArAGhgAOAJxACAutnZQKAgKYgAZvH4BrWRUUxpczXDEwAPjCkz5SiqvWbtZmBTEBKJAD5UxGDCGyoAVyFgYLm52YVExEiJiIA
